### ISSUE TITLE
Class based views

### ISSUE THREAD

**Issue by barseghyanartur** [barseghyanartur/django-fobi#147]
Move to class based views

**Comment by jstacoder** [comment]
Has any progress been made here yet? If not, I have been able to convert `create_form_wizard_entry` and `edit_form_wizard_entry` into classes, and I’m working on the others, I could try integrating them into the repo? Maybe as a separate module for now? Outside of fobi/views.py, maybe in like fobi/generic/views.py and maybe fobi/views/*.py, thoughts?

**Comment by barseghyanartur** [comment]
@jstacoder:

Any help is highly appreciated.

I guess the best way would be to create CBV next to FBV and when everything is ready, just switch.

I have prepared that (views module has not a sub-module).

**Comment by jstacoder** [comment]
Ok, I see you made a place for them, awesome. I’ll start adding pull requests after the weekend.

**Comment by jstacoder** [comment]
The error happens on line 796 of fobi/base.py

**Comment by barseghyanartur** [comment]
I'll check it asap.

**Comment by jstacoder** [comment]
here is a [pastebin](https://pastebin.com/b11stu3G) of the class i currently have, maybe it will give clues as to where im going wrong. [`AddFormElementEntryView`](https://pastebin.com/b11stu3G)

**Comment by jstacoder** [comment]
n/m managed to figure out where i was going off in the flow of things

**Comment by jstacoder** [comment]
@barseghyanartur  i was thinking of a clean way to enable using either the class based views vs functions, and i love the Theme class, so i decided that a ClassBasedViewsTheme would be perfect, but then i realized i would have a ton of templates to override to get the urls pointing the right way, then i realized if i just tuck the url endpoints into the theme, then it fully encapsulates the floating data in the templates, ie: no more `'strings'` hardcoded into the templates, then instead of `{% url 'fobi.create_form_entry' %}` it could be `{% url fobi_theme.create_form_entry %}` which gives the power to the theme to provide the urls to route to as well, so then i just need to override the endpoints in the new theme and not a single template to get everything pointing to the classes inside of the classes, although it does mean a lot of small template updates, i hope you dont mind?

**Comment by barseghyanartur** [comment]
Well, look, once class based views are stable, we can make the switch. Meanwhile, it's easier to have two separate url configs (ans since that's something which is specified in settings, we can run tests for both). All existing templates would not have to change.

**Comment by barseghyanartur** [comment]
@jstacoder:

Hey, thanks a lot! I appreciate your work.

There is one remark I have regarding the switch from function based views to class based views.

- The class-based-views shall reflect 100% functionality of the function-based-views.
- The class-based-views shall be totally independant from function-based-views. That means, that even if function-based-views contains certain class-based-views, those class-based-views shall be copied into class-based-views as is, so that we can make a swith later.
- Same shall be done to the URLs. Note, that in order to make things flexible, edit-views and separated from view-views. So, just preserve them separate (I have just moved them into a separate sub-modules).

Thus, finally, what we're gonna have is:

- class-based-views 100% equal in terms of functionality with the function-based-views.
- URLs for class-based-views 100% equal in terms of functionality with the function-based-views URLs.

As I mentioned earlier, in the tests we will be testing both. 

As of the roadmap:

- Version 0.13 we will introduce (at first, experimental) class-based-views.
- In the release notes of 0.13 there will be a note, that function based views will be deprecated in 0.14.
- Throughout the entire 0.13.x there will be patches/fixes applied, whenever necessary. Once it feels stable, 0.14 will be relaesed.

**Comment by barseghyanartur** [comment]
P. S.

I have just manually reverted some backwards incompatible changes made to the templates.

**Comment by jstacoder** [comment]
Yep that’s the goal , 100% functional compatibility, and so far so good I believe, plus at the moment with things mixed between classes and functions it still works perfect. And yesterday I looked over the views left, looks like about 7 down 20 to go.

**Comment by barseghyanartur** [comment]
@jstacoder:

Hey, what's the state of this? I want to make a new release one day soon. Do you think it's testable already?

**Comment by jstacoder** [comment]
At my company we have been using the classes that have been written so far quite successfully, so things are definitely working enough for tests, I was looking at what’s left the other day, and it’s mainly the handler stuff remaining, that’s something we aren’t going to make dynamic (users will create forms, but we assign default handlers), I’ve been meaning to get to them

**Comment by barseghyanartur** [comment]
@jstacoder:

I'll add a dedicated settings file to test CBV functionality and prepare an 014 release.

**Comment by barseghyanartur** [comment]
@jstacoder:

I"ve added a [wiki page for the transition](https://github.com/barseghyanartur/django-fobi/wiki/Function-based-views-to-class-based-views-transition).

The current plan:

- Import the missing views from function-based-views modules into class-based-views modules. Change things as there's time.
- Version 0.14.x will introduce a non-breaking change - the optional class-based-views. Devs will be encouraged to test them. Old function-based-views would be still set as default.
- Version 0.15 will be released somewhere in the end of 2018. The switch to class-based-views will be turned on by default. 
- Old function-based-views will remain until at least version 0.16.

**Comment by barseghyanartur** [comment]
Class based views are currently being written from scratch. WIP in branch #147.

**Comment by barseghyanartur** [comment]
Version 0.19 has just been released. It should be considered as the first step. Note, that form wizard views are not yet class-based. Will be handled in a separate issue.
